src/recordlinker/linkage/models.py

```diff
@@ -47,3 +44,48 @@ class BlockingValue(Base):
     patient_id: orm.Mapped[int] = orm.mapped_column(ForeignKey("mpi_patient.id"))
     blockingkey_id: orm.Mapped[int] = orm.mapped_column(ForeignKey("mpi_blocking_key.id"))
     value: orm.Mapped[str] = orm.mapped_column(String(50), index=True)
+
+class Algorithm(Base):
+    __tablename__ = "algorithm"
+
+    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
+    is_default: orm.Mapped[bool] = orm.mapped_column(default=False)
+    label: orm.Mapped[str] = orm.mapped_column(String(255), unique=True)
+    description: orm.Mapped[str]
+
+def check_only_one_default(connection, target):
+    """
+    Check if there is already a default algorithm before inserting or updating.
+
+    Called before an insert or update operation on the
+    Algorithm table. If the `is_default` attribute of the target object is
+    set to True, it checks the database to ensure that no other algorithm
+    is marked as default. If another default algorithm exists, an exception
+    is raised to prevent the operation.
+
+    Parameters:
+    connection: The database connection being used for the operation.
+    target: The instance of the Algorithm class being inserted or updated.
+    
+    Raises:
+    Exception: If another algorithm is already marked as default.
+    """
+     
+    if target.is_default:
+        existing_default = connection.execute("SELECT COUNT(*) FROM algorithm WHERE is_default = TRUE").scalar()
```

Comment:
Eventually, we'll be adding an optional configuration settings to the project for database table prefixes via a [declared_attr](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.declared_attr) or something like that.

We know that NBS would like to have these tables live in an existing Microsoft SQL Server, and other STLTs may want to do the same down the road.  This opens up the potential for table name clashes, as the NBS team may already have a table named "algorithm" on their server.

If we stick to using the ORM, we can make this dynamic and allow the users to add/change the default prefix if they wish so, to prevent any table name clashes.

Could you rewrite this query using the ORM?

Reply:
updated